compiler/src/dotty/tools/dotc/typer/EtaExpansion.scala

```diff
@@ -223,6 +223,9 @@ object EtaExpansion extends LiftImpure {
    */
   def etaExpand(tree: Tree, mt: MethodType, xarity: Int)(using Context): untpd.Tree = {
     import untpd._
+    if ctx.phase == Phases.inliningPhase then
+      report.error(em"Cannot automaticaly eta-expand calls in macros: $tree", tree.srcPos)
+      return EmptyTree
```

Comment:
Can we do the check from the place where `etaExpand` is called?

Reply:
Actually, we can catch it as soon as it is created by the macro and give the stack trace to the user.

Reply:
Now it will fail under `ycheck` with

```scala
24 |    X.process[Future,Unit]{ // error
   |    ^
   |Exception occurred while executing macro expansion.
   |java.lang.AssertionError: assertion failed: Reference to a method must be eta-expanded before it is used as an expression: (
   |  {
   |    def $anonfun(x: concurrent.Future[String]): String = 
   |      {
   |        x.await[concurrent.Future, String](x)(x.FutureAsyncMonad)
   |      }
   |    closure($anonfun:Conversion[concurrent.Future[String], String])
   |  }
   |:Conversion[concurrent.Future[String], String]).apply
   |    at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
   |    at scala.quoted.runtime.impl.QuotesImpl$reflect$.scala$quoted$runtime$impl$QuotesImpl$reflect$$$yCheckValidExpr(QuotesImpl.scala:2884)
   |    at scala.quoted.runtime.impl.QuotesImpl$reflect$Apply$.yCheckArgs$$anonfun$1(QuotesImpl.scala:603)
   |    at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
   |    at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
   |    at scala.collection.immutable.List.foreach(List.scala:333)
   |    at scala.quoted.runtime.impl.QuotesImpl$reflect$Apply$.yCheckArgs(QuotesImpl.scala:603)
   |    at scala.quoted.runtime.impl.QuotesImpl$reflect$Apply$.apply(QuotesImpl.scala:595)
   |    at scala.quoted.runtime.impl.QuotesImpl$reflect$Apply$.apply(QuotesImpl.scala:594)
   |    at x.X$.processTree(Macro_1.scala:50)
   |    at x.X$.processTree(Macro_1.scala:45)
   |    at x.X$.processTree(Macro_1.scala:42)
   |    at x.X$.processTree(Macro_1.scala:39)
   |    at x.X$.processImpl(Macro_1.scala:32)
```
This lets the user know that he created the wrong tree at `Macro_1.scala:50`